File: mac/tools/crosscheck/streamhelp.swift
import Foundation

var out: [String] = []
for line in StreamHelp.before { out.append("before|\(line)") }
for line in StreamHelp.after { out.append("after|\(line)") }
out.append("order|[" + StreamHelp.order.map { "'\($0)'" }.joined(separator: ", ") + "]")
for p in StreamHelp.order + ["nonsense"] {
    for (heading, items) in StreamHelp.stepsFor(p) {
        out.append("head|\(p)|\(heading)")
        for (i, item) in items.enumerated() {
            out.append("  item|\(p)|\(heading)|\(i)|\(item)")
        }
    }
    // Python's splitlines() drops a single trailing newline rather than
    // yielding an empty last element, so match that.
    var textLines = StreamHelp.asText(p)
        .split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
    if textLines.last == "" { textLines.removeLast() }
    out.append("text|\(p)|\(textLines.count) lines")
    for (n, line) in textLines.enumerated() { out.append("  line|\(p)|\(n)|\(line)") }
}
for line in StreamHelp.picture { out.append("picture|\(line)") }
for line in StreamHelp.framing { out.append("framing|\(line)") }
for t in StreamHelp.trouble { out.append("trouble|\(t.what)|\(t.fix)") }
let whole = StreamHelp.everything()
let wholeLines = whole.split(separator: "\n", omittingEmptySubsequences: false)
    .map(String.init)
var counted = wholeLines
if counted.last == "" { counted.removeLast() }
out.append("everything|\(counted.count) lines")
for (n, line) in counted.enumerated() { out.append("  all|\(n)|\(line)") }
print(out.joined(separator: "\n"))

File: mac/tools/crosscheck/streamhelp.py
# The repository root is on the path already, put there by cross_check.py.
from dropdeck import streamhelp as S

out = []
for line in S.BEFORE:
    out.append("before|%s" % line)
for line in S.AFTER:
    out.append("after|%s" % line)
out.append("order|%s" % list(S.ORDER))
for p in S.ORDER + ("nonsense",):
    for heading, items in S.steps_for(p):
        out.append("head|%s|%s" % (p, heading))
        for i, item in enumerate(items):
            out.append("  item|%s|%s|%d|%s" % (p, heading, i, item))
    out.append("text|%s|%d lines" % (p, len(S.as_text(p).splitlines())))
    for n, line in enumerate(S.as_text(p).splitlines()):
        out.append("  line|%s|%d|%s" % (p, n, line))
for line in S.PICTURE:
    out.append("picture|%s" % line)
for line in S.FRAMING:
    out.append("framing|%s" % line)
for what, fix in S.TROUBLE:
    out.append("trouble|%s|%s" % (what, fix))
whole = S.everything()
out.append("everything|%d lines" % len(whole.splitlines()))
for n, line in enumerate(whole.splitlines()):
    out.append("  all|%d|%s" % (n, line))
print("\n".join(out))
